File: frontend/src/types/timeline.ts
export type TimelineEventType = "metric" | "log" | "deployment";

export interface TimelineEvent {
  id: string;
  timestamp: string;
  event_type: TimelineEventType;
  service_id: string;
  title: string;
  description: string | null;
  source_id: string;
  severity: string | null;
  metadata: Record<string, unknown>;
}

export interface TimelineResponse {
  incident_id: string;
  start_time: string;
  end_time: string;
  events: TimelineEvent[];
}

File: backend/app/schemas/timeline.py
from datetime import datetime
from typing import Any, Literal
from uuid import UUID

from pydantic import BaseModel, Field


TimelineEventType = Literal["metric", "log", "deployment"]


class TimelineEvent(BaseModel):
    id: UUID
    timestamp: datetime
    event_type: TimelineEventType
    service_id: UUID
    title: str
    description: str | None = None
    source_id: UUID
    severity: str | None = None
    metadata: dict[str, Any] = Field(default_factory=dict)


class TimelineResponse(BaseModel):
    incident_id: UUID
    start_time: datetime
    end_time: datetime
    events: list[TimelineEvent]